lib/terraform/token_source.go

```diff
@@ -0,0 +1,65 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package terraform
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/gravitational/trace"
+)
+
+type envGetter func(key string) string
+
+// IDTokenSource allows a Terraform Workload ID token to be fetched whilst
+// within a Terraform Cloud execution.
+type IDTokenSource struct {
+	audienceTag string
+
+	getEnv envGetter
+}
+
+func (its *IDTokenSource) GetIDToken() (string, error) {
+	name := "TFC_WORKLOAD_IDENTITY_TOKEN"
+	if its.audienceTag != "" {
+		name = fmt.Sprintf("TFC_WORKLOAD_IDENTITY_TOKEN_%s", strings.ToUpper(its.audienceTag))
```

Comment:
Wow - had no clue they put the audience into the name of the env var - that's super awkward!

Reply:
Yep, it's a weird one. Well, not quite in the _name_ of the var - the audience tag is an optional parameter in the var name. The audience is the var value.

The rule is more or less: whenever TFC sees a var matching `TFC_WORKLOAD_IDENTITY_AUDIENCE(?P<tag>_\w+)?` during a run, it uses the value as the audience, then writes the JWT into `$TFC_WORKLOAD_IDENTITY_TOKEN$tag`.

The tag is meant to let you issue more than one token, since otherwise you could only define one audience. More docs here: https://developer.hashicorp.com/terraform/enterprise/workspaces/dynamic-provider-credentials/specifying-multiple-configurations

---

(Naturally, none of this applies to their own providers which use some slightly different magic to write the same content into different env vars 🤷)